### PR TITLE
#524 Apply Quick Search as New Design

### PR DESCRIPTION
**Pull request by dab246**
- Result:

https://user-images.githubusercontent.com/80730648/169964503-fba85087-224f-4cdf-9897-01109d233991.mov

https://user-images.githubusercontent.com/80730648/169964758-55a8a635-1079-4eae-b169-ec3e2d0263c4.mp4

### CODE REVIEW

File: lib/features/email/presentation/email_controller.dart [hunk 1]
```diff
@@ -614,6 +614,7 @@ class EmailController extends BaseController {
 
   void backToThreadView(BuildContext context) {
     mailboxDashBoardController.clearSelectedEmail();
+    mailboxDashBoardController.disableSearch();
```
Comment (hoangdat) [inline lib/features/email/presentation/email_controller.dart:617]: in the aspect of UX, IMO no problem when user want to search when he stays in `EmailView`

File: lib/features/mailbox_dashboard/domain/usecases/quick_search_email_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import 'package:dartz/dartz.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/filter/filter.dart';
+import 'package:jmap_dart_client/jmap/core/properties/properties.dart';
+import 'package:jmap_dart_client/jmap/core/unsigned_int.dart';
+import 'package:jmap_dart_client/jmap/core/sort/comparator.dart';
+import 'package:tmail_ui_user/features/mailbox_dashboard/domain/state/quick_search_email_state.dart';
+import 'package:tmail_ui_user/features/thread/domain/repository/thread_repository.dart';
+
+class QuickSearchEmailInteractor {
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/domain/usecases/quick_search_email_interactor.dart:13]: why not re-use the interactor for search which implemented before?

File: lib/features/cleanup/data/datasource_impl/cleanup_datasource_impl.dart [hunk 3]
```diff
@@ -22,19 +22,9 @@ class CleanupDataSourceImpl extends CleanupDataSource {
...
-      await sharedPreferences.setString(CleanupConstant.keyTimeCleanupEmail, lastTime.toIso8601String());
-    }).catchError((error) {
-      throw error;
-    });
-  }
-
-  @override
-  Future<DateTime?> getTimeCleanEmail() {
-    return Future.sync(() async {
-      final lastTime = sharedPreferences.getString(CleanupConstant.keyTimeCleanupEmail);
```
[the remark this answers is not in the document]
Comment (dab246) [reply lib/features/cleanup/data/datasource_impl/cleanup_datasource_impl.dart:36]: Yes, cleaning up email is not affected. It still plays out but the original script.

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart [hunk 4]
```diff
@@ -396,24 +393,170 @@ class MailboxDashBoardView extends GetWidget<MailboxDashBoardController> with Ne
...
+              onTap: () {
+                controller.disableSearch();
+                controller.dispatchAction(CancelSelectionAllEmailAction());
+              }),
+          Expanded(child: Container(
+            decoration: const BoxDecoration(
+              borderRadius: BorderRadius.all(Radius.circular(10)),
+              color: AppColor.colorBgSearchBar),
+            height: 45,
+            child: TypeAheadFormFieldQuickSearch<PresentationEmail, RecentSearch>(
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:412]: How about move your modification in `flutter_typeahead` into `tmail-flutter`? IMO, it should be good for maintaining

File: lib/features/thread/presentation/thread_controller.dart [hunk 5]
```diff
@@ -836,15 +831,56 @@ class ThreadController extends BaseController {
...
 
     if (_accountId != null && searchQuery != null) {
+      UTCDate? afterTime;
+      bool? hasAttachment;
+      String? fromAddress;
+      String? keywordUnread;
+      String? keywordStarred;
+
+      if (mailboxDashBoardController.listFilterQuickSearch.contains(QuickSearchFilter.last7Days)
+          && mailboxDashBoardController.emailReceiveTimeType.value != null) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:849]: Move to a method, it is a good practice for readable code without comment

File: lib/features/thread/presentation/thread_controller.dart [hunk 6]
```diff
@@ -836,15 +831,56 @@ class ThreadController extends BaseController {
...
+        afterTime = mailboxDashBoardController.emailReceiveTimeType.value!.toUTCDate();
+      }
+
+      if (mailboxDashBoardController.listFilterQuickSearch.contains(QuickSearchFilter.hasAttachment)
+        || mailboxDashBoardController.filterMessageOption.value == FilterMessageOption.attachments) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:854]: idem

File: lib/features/thread/presentation/thread_controller.dart [hunk 7]
```diff
@@ -836,15 +831,56 @@ class ThreadController extends BaseController {
...
+        hasAttachment = true;
+      }
+
+      if (mailboxDashBoardController.listFilterQuickSearch.contains(QuickSearchFilter.fromMe)) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:858]: idem

File: lib/features/thread/presentation/thread_controller.dart [hunk 8]
```diff
@@ -836,15 +831,56 @@ class ThreadController extends BaseController {
...
+        fromAddress = mailboxDashBoardController.userProfile.value?.email;
+      }
+
+      if (mailboxDashBoardController.filterMessageOption.value == FilterMessageOption.unread) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:862]: idem

File: lib/features/thread/presentation/thread_controller.dart [hunk 9]
```diff
@@ -836,15 +831,56 @@ class ThreadController extends BaseController {
...
+        keywordUnread = KeyWordIdentifier.emailSeen.value;
+      }
+
+      if (mailboxDashBoardController.filterMessageOption.value == FilterMessageOption.starred) {
```
Comment (hoangdat) [inline lib/features/thread/presentation/thread_controller.dart:866]: idem

**Comment by hoangdat** [comment]
Base on the demo, IMO, we should keep the `threadView` as normal when search is active.

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:412]: That is easy. But do you want to do that? I think we should put it in a separate library so everyone can use it.

Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_view_web.dart:412]: the search widget is too specific for our app only, so please move that to our code base.

**Comment by dab246** [comment]
> Base on the demo, IMO, we should keep the `threadView` as normal when search is active.



https://user-images.githubusercontent.com/80730648/170174636-e7d34fad-4d54-4f5f-9aa1-a3d41f0af98e.mov

File: pubspec.yaml [hunk 10]
```diff
@@ -64,9 +64,6 @@ dependencies:
   # flutter_dotenv
   flutter_dotenv: 5.0.2
 
-  # flutter_typeahead
-  flutter_typeahead: 3.1.1
```
Comment (hoangdat) [inline pubspec.yaml:68]: we don't need it anymore? weird. I think, we only add the customization in `tmail-flutter`, all the other base of `flutter_typeahead` we still use from the upstream lib

Comment (dab246) [reply pubspec.yaml:68]: Currently, we don't use it anymore

**Review by hoangdat — APPROVED** [review]

**Pull request merged**